### ISSUE TITLE
Memory leak when calling gl.GLMeshItem.setMeshData()

### ISSUE THREAD

**Issue by solarjoe** [pyqtgraph/pyqtgraph#825]
pyqtgraph 0.10.0
Python 3.5
Qt 5.6.2

I came across a issue and suppose it is a memory leak, maybe similar to https://github.com/pyqtgraph/pyqtgraph/issues/103

When I update vertices of a `gl.GLMeshItem` using its method `.setMeshData()` I loose memory each time it updates. You can reduce `n_tri` for a less severe example ;)

A working example is 

```
# -*- coding: utf-8 -*-
from pyqtgraph.Qt import QtGui, QtCore

import numpy as np
import pyqtgraph as pg

import pyqtgraph.opengl as gl

app = QtGui.QApplication([])
mw = QtGui.QMainWindow()
mw.setWindowTitle('pyqtgraph example: PlotWidget')
mw.resize(800,800)
cw = QtGui.QWidget()
mw.setCentralWidget(cw)

l = QtGui.QGridLayout()

#l.setRowMinimumHeight(0, 400)
cw.setLayout(l)

w = gl.GLViewWidget()
w.setBackgroundColor(0.9)
w.setCameraPosition(distance=3)

l.addWidget(w, 0, 0, 1, 1)

sp = w.sizePolicy()
w.setSizePolicy(QtGui.QSizePolicy(QtGui.QSizePolicy.Expanding, QtGui.QSizePolicy.Expanding))

mw.show()

n_tri = 10000

vertices = np.random.uniform(size=(n_tri, 3, 3))

colors = np.random.uniform(size=(n_tri, 3, 4))

m1 = gl.GLMeshItem(vertexes=vertices, faces=None, vertexColors=colors, smooth=False,
                   drawEdges=False, drawFaces=True, edgeColor=(0, 1, 0, 1))

w.addItem(m1)

def update():

    vertices = np.random.uniform(size=(n_tri, 3, 3))

    m1.setMeshData(vertexes=vertices, faces=None, vertexColors=colors, smooth=False,
                   drawEdges=False, drawFaces=True, edgeColor=(0, 1, 0, 1))

t2 = QtCore.QTimer()
t2.timeout.connect(update)
t2.start(500)

if __name__ == '__main__':
    import sys
    if (sys.flags.interactive != 1) or not hasattr(QtCore, 'PYQT_VERSION'):
        QtGui.QApplication.instance().exec_()
```

Any ideas what causes this and if there is a workaround or a fix?

**Comment by solarjoe** [comment]
I just saw that I did not handle "indexed" vertices correctly and hoped that this was the reason, but it was not. Fixed example.

**Comment by solarjoe** [comment]
Ok, I guess I am getting closer. (This)[https://sourceforge.net/p/pyopengl/mailman/message/6175890/] got me on the right track

> Just changing glColorPointerf(self.vertexColors) by 
> glColorPointerd(self.vertexColors) solved my problem.

I looked at the source of (`GLMeshItem.paint()`)[https://github.com/pyqtgraph/pyqtgraph/blob/develop/pyqtgraph/opengl/items/GLMeshItem.py#L179] and saw that it is using `glVertexPointerf`.

Usually Numpy arrays are doubles, not floats. for testing I changed the related functions in `GLMeshItem.paint()`

```
glVertexPointerf
glNormalPointerf
glColor4f
glColorPointerf
```
to their double versions

```
glVertexPointerd
glNormalPointerd
glColor4d
glColorPointerd
```

and the memory leak is gone. 

Does that make any sense? Why could this be a problem? Will PyQtGraph only free half of the memory?

**Comment by solarjoe** [comment]
Another fix is to add `.astype(np.float32)` to the input, but this won't fix it for the normals as there is the following line in https://github.com/pyqtgraph/pyqtgraph/blob/develop/pyqtgraph/opengl/MeshData.py#L211

`self._vertexNormals = np.empty(self._vertexes.shape, dtype=float)`

so this is always a Python float, which is a double.

**Comment by j9ac9k** [comment]
@pijyoi  pointed out that this was likely addressed in

**Comment by pijyoi** [comment]
Don't think it was addressed, although it seems to be a similar issue. #1800's issue is when using faces indices. This issue is when using vertex (float) coords.

**Comment by j9ac9k** [comment]
> Don't think it was addressed, although it seems to be a similar issue. #1800's issue is when using faces indices. This issue is when using vertex (float) coords.

Thanks, I misread! Re-Opened

**Comment by pijyoi** [comment]
Hello @solarjoe , I tried out your script on (Windows 10 21H2 + Ubuntu 20.04 / PyQt6 6.2.2 / numpy 1.21.4 / PyOpenGL 3.1.5 / pyqtgraph master) and was not able to observe any growth in memory usage. I even reduced the timer to 50ms.

Now even in issue #1783, I was not able to observe a memory leak, but the user raising the issue was able to observe the leak on their machine. So the observations on my machines may not count.

Given that almost 3 years have passed since you raised this issue, would it be possible for you to check if the issue still persists in current versions of Qt and pyqtgraph?

I agree with your comments in https://github.com/pyqtgraph/pyqtgraph/issues/825#issuecomment-461729867. pyqtgraph is triggering PyOpenGL's automatic data type conversion.

**Comment by solarjoe** [comment]
I could not test with PyQt6 because there are dependency conflicts for the wheel or whatever.
But I still see the issue with the latest pyqtgraph and PyQt5. I also reduced the timer to 50 ms.
The right graph is the memory consumption with the script running for about 2 minutes, then cancelled. 

![grafik](https://user-images.githubusercontent.com/10460113/144702064-e0041abd-1b69-48cd-9109-8ddfcc361ab7.png)

Linux Mint 19.3, 64-bit
Python 3.7

New environment, pip list:

numpy               1.21.4
pip                 21.3.1
pkg_resources       0.0.0
PyOpenGL            3.1.5
PyOpenGL-accelerate 3.1.5
PyQt5               5.15.6
PyQt5-Qt5           5.15.2
PyQt5-sip           12.9.0
pyqtgraph           0.12.3
scipy               1.7.3
setuptools          59.4.0
wheel               0.37.0

At the moment I don't have time to apply the workaround I mentioned in https://github.com/pyqtgraph/pyqtgraph/issues/825#issuecomment-461729867 but if you can provide some monkeypatch to override the methods that I just have to add to the script I will run it.

**Comment by pijyoi** [comment]
I have created PR #2131 which avoids PyOpenGL array automatic conversion. Would you be able to test it out to see if stops the memory leak?

**Comment by pijyoi** [comment]
Okay, I have finally been able to reproduce the memory leak in Ubuntu 20.04. The memory leak occurs when ```pyopengl-accelerate``` is installed. (I don't normally install ```pyopengl-accelerate```)

I can't reproduce it on Windows with ```pyopengl-accelerate``` installed. I get the following warning message:
```Unable to load numpy_formathandler accelerator from OpenGL_accelerate```,
which I suppose is very relevant to the memory leak.

As I am able to reproduce the memory leak now, I was able to test PR #2131 and can confirm that with it, the memory leak no longer occurs.
